### PR TITLE
Reject disallowed store receipts (bug 1135261)

### PR DESCRIPTION
**Pull request by kumar303**
@mstriemer r?

### CODE REVIEW

**Comment by kumar303** [comment]
or @muffinresearch r?

File: tests/test-validate-app-receipt.js [hunk 1]
```diff
@@ -75,6 +75,63 @@ describe('fxpay.validateAppReceipt()', function() {
     validator.finish();
   });
 
+  it('fails when receipt was issued by a disallowed store', function(done) {
+    // Disallow all other stores:
+    helper.appSelf.manifest.installs_allowed_from = [
```
Comment (mstriemer) [inline tests/test-validate-app-receipt.js]: Does this get set in a `beforeEach`? It looks like this change could leak into the next test.

File: lib/fxpay/receipts.js [hunk 2]
```diff
@@ -142,6 +142,28 @@
...
+      var manifest = settings.appSelf.manifest;
+      var allowAny;
+      if (!manifest.installs_allowed_from) {
+        // This is an unlikely case but let's guess that it implies "*".
+        allowAny = true;
+      } else {
+        allowAny = manifest.installs_allowed_from.indexOf('*') !== -1;
+      }
+      if (allowAny) {
+        settings.log.warn('your paid app manifest specifies ' +
```
Comment (mstriemer) [inline lib/fxpay/receipts.js]: Should this new formatted like a sentence?

**Comment by mstriemer** [comment]
Looks good as long as the test

File: tests/test-validate-app-receipt.js (hunk 1 above)
Comment (kumar303) [inline tests/test-validate-app-receipt.js]: Yeah, it gets set before each test here https://github.com/mozilla/fxpay/pull/75/files#diff-d5ebdbfb88cc7626c6e2f1d9f32e6d5aR161

File: lib/fxpay/receipts.js (hunk 2 above)
Comment (kumar303) [inline lib/fxpay/receipts.js]: Could you rephrase the question?

Comment (mstriemer) [inline lib/fxpay/receipts.js]: Start with a capital and end with a period.

**Comment by kumar303** [comment]
mmm, I refactored all code that relied on initialization of appSelf so that it gets the value more lazily but that patch became sort of epic. It removes some async madness from the tests which is nice though.

**Comment by mstriemer** [comment]
That grew significantly. I'm not sure I can comment on everything that was changed but it seems reasonable through a once-over.

What was the reasoning for the change, exactly? I guess previously `appSelf` was retrieved on startup and if the `receipts.all()` was used too soon it would blow up? Does this remove the need for that `initialized` event that would happen?

**Comment by kumar303** [comment]
It is mostly a refactor job and a removal of some TODOs that are related to how appSelf is accessed. Yeah, whoops, it sort of became a big patch. The initialization was very in-app specific and now that we have app receipt validation it doesn't make sense. What this refactor does is push access to appSelf down to the code that actually needs it so that initialization errors are more lazy and occur at the point of access.

File: lib/fxpay/adapter.js [hunk 3]
```diff
@@ -37,28 +37,39 @@
...
+
+      receipts.all(function(error, allReceipts) {
+        if (error) {
+          return callback(error);
+        }
+        var numReceipts = 0;
+
+        allReceipts.forEach(function(receipt) {
+          settings.log.info('Installed receipt: ' + receipt);
+          numReceipts++;
```
Comment (mstriemer) [inline lib/fxpay/adapter.js]: I see that this didn't change but it looks like it could use `allReceipts.length` when logging instead of counting.

Comment (kumar303) [inline lib/fxpay/adapter.js]: good call

File: tests/test-get-products.js [hunk 4]
```diff
@@ -90,23 +90,4 @@ describe('fxpay.getProducts()', function() {
 
     helper.server.respond();
   });
-
-  it('should pass through init errors', function (done) {
-    // Trigger an init error:
-    helper.appSelf.error = {name: 'INVALID_MANIFEST'};
```
Comment (mstriemer) [inline tests/test-get-products.js]: This test still sounds valid to me, just that the setup would be different.

**Comment by mstriemer** [comment]
Looks good. r+wc

File: tests/test-get-products.js (hunk 4 above)
Comment (kumar303) [inline tests/test-get-products.js]: actually, this was interesting because ``products.all()`` doesn't need ``appSelf`` at all so it doesn't need to pass through any initialization errors. That made me realize lazy access to appSelf solves a lot of problems. The concept of initialization always bothered me and I think you also had some early review comments about how up front initialization was clunky :)

**Comment by kumar303** [comment]
Thanks for taking a second look. I'm going to do a quick Marketplace test locally with it to make sure things are still working then merge it in.

**Comment by kumar303** [comment]
Ugh, trying to make immediate callbacks async with setTimeout is blowing up the test suite and I can't track down exactly why. Looks like there may be a bug in there somewhere. However, I think moving to promises will make everything async by default.

**Comment by kumar303** [comment]
Bah. My local Marketplace is hosed but this is working in the example app.
